src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1467,27 +1467,40 @@ internal void ClearTransaction()
 
         internal void PerformUserCancellation()
         {
-            _userCancellationRequested = true;
-
-            if (AttemptPostgresCancellation && SupportsPostgresCancellation)
+            lock (CancelLock)
```

Comment:
Agree:  on a first sight if `PerformPostgresCancellation` has this then `PerformUserCancellation` should have as well, although there is a question:
- wasn't a lockless `PerformUserCancellation` implementation a part of the design, not a bug? `lock (this)` protects the state of a connector instance. To avoid the race without having to lock `PerformUserCancellation`: `PerformPostgresCancellation` creates another connector instance that it then cancels. This logic based on operating on a cloned connector instance should've been sufficient to protect from the race condition and the damage to the original connector state. By adding another `lock` we may hide some other bug at a potential cost of the decreased performance.

Reply:
`PerformPostgresCancellation` creates another connector instance because that's how cancellation works for PG ([link](https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.9)). It's definitely not to avoid locking.
And no, lockless `PerformUserCancellation` is a bug, as:
a) `lock` around `CancelLock` is required, as we have to be sure while sending a query what we're not concurrently cancelling previous query
https://github.com/npgsql/npgsql/blob/d413846ad999768b91467219d9f8b5632147e923/src/Npgsql/NpgsqlCommand.cs#L1220-L1223
b) `lock` around `this` is required, because `PerformUserCancellation` works with disposable resources (`ReadBuffer`). But, I do hope by dealing with #3045 we can rewrite this part and remove locking.
While there will be some performance degradation, it's very limited (only for the cancelled queries while the connector is being broken/closed).

Reply:
I could not find `PerformPostgresCancellation` invoked any other way but through `PerformUserCancellation`. There is one other place in NpgsqlReader that is never executed in the current code base, so `PerformPostgresCancellation` can be safely made `private` and `CancelLock` only left for `PerformUserCancellation` own use.

Reply:
https://github.com/npgsql/npgsql/blob/d413846ad999768b91467219d9f8b5632147e923/src/Npgsql/NpgsqlReadBuffer.cs#L228

Reply:
My bad.